File: _raw/example1/src/main/java/com/example/grpc/echo/unary/UnaryEchoAsynchronousClient.java
package com.example.grpc.echo.unary;

import io.grpc.Grpc;
import io.grpc.InsecureChannelCredentials;
import io.grpc.ManagedChannel;
import io.grpc.Status;
import com.example.grpc.echo.EchoRequest;
import com.example.grpc.echo.EchoResponse;
import com.example.grpc.echo.EchoServiceGrpc;
import com.example.grpc.echo.Logging;
import io.grpc.stub.StreamObserver;

import java.util.concurrent.TimeUnit;
import java.util.logging.Logger;

public class UnaryEchoAsynchronousClient {

    private static final Logger logger = Logger.getLogger(UnaryEchoAsynchronousClient.class.getName());

    public static void main(String[] args) throws Exception {
        Logging.init();

        ManagedChannel channel = Grpc.newChannelBuilder("localhost:50051", InsecureChannelCredentials.create()).build();

        EchoServiceGrpc.EchoServiceStub asyncStub = EchoServiceGrpc.newStub(channel);
        EchoRequest request = EchoRequest.newBuilder().setMessage("world").build();
        asyncStub.unaryEcho(request, new StreamObserver<EchoResponse>() {
            @Override
            public void onNext(EchoResponse response) {
                logger.info("next: " + response.getMessage());
            }

            @Override
            public void onError(Throwable t) {
                logger.warning("client error: {0}" + Status.fromThrowable(t));

                logger.info("error: " + t);
            }

            @Override
            public void onCompleted() {
                logger.info("completed");
            }
        });

        channel.shutdown().awaitTermination(10, TimeUnit.SECONDS);
    }
}

File: _raw/kotlin/src/main/java/com/example/grpc/echo/unary/UnaryEchoAsynchronousClient.kt
package com.example.grpc.echo.unary

import com.example.grpc.echo.EchoRequest
import com.example.grpc.echo.EchoResponse
import com.example.grpc.echo.EchoServiceGrpc
import com.example.grpc.echo.Logging
import io.grpc.Grpc
import io.grpc.InsecureChannelCredentials
import io.grpc.Status
import io.grpc.stub.StreamObserver
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.logging.Logger

object UnaryEchoAsynchronousClient {
    private val logger: Logger = Logger.getLogger(UnaryEchoAsynchronousClient::class.java.getName())

    @kotlin.Throws(Exception::class)
    @kotlin.jvm.JvmStatic
    fun main(args: Array<String>) {
        Logging.init()

        val channel = Grpc.newChannelBuilder("localhost:50051", InsecureChannelCredentials.create()).build()

        val asyncStub = EchoServiceGrpc.newStub(channel)
        val request = EchoRequest.newBuilder().setMessage("world").build()

        val latch = CountDownLatch(1)
        asyncStub.unaryEcho(request, object : StreamObserver<EchoResponse?> {
            override fun onNext(response: EchoResponse) {
                logger.info("next: " + response.getMessage())
            }

            override fun onError(t: Throwable) {
                logger.warning("error: " + Status.fromThrowable(t))
                latch.countDown()
            }

            override fun onCompleted() {
                logger.info("completed")
                latch.countDown()
            }
        })

        latch.await()
        channel.shutdown().awaitTermination(10, TimeUnit.SECONDS)
    }
}
